internal/addr.go

```diff
@@ -44,8 +43,21 @@ func ParseAddr(addr string) (string, int, error) {
 	if host == "" || strings.TrimSpace(host) == "" {
 		host = defaultHost
 	}
-	if portInt < 0 { // port number should be more than 0
+	if portInt < 0 {
+		// the port number must be positive value
 		return "", 0, fmt.Errorf("invalid port number: '%d'", portInt)
 	}
 	return host, portInt, nil
 }
+
+func hasPort(addr string) bool {
+	i := strings.LastIndex(addr, ":")
+	if i == -1 {
+		return false
+	}
+	if i == len(addr)-1 {
+		return true
+	}
```

Comment:
Is this check necessary?

Reply:
It is necessary as we get `i+1` as an index of `addr` at the rest

Reply:
That condition makes the function return `true` for `192.168.1.1:`. Is that what was intended?

Reply:
Returning `true` for something like `192.168.1.1` is misleading, and may be an indicator that `hasPort` isn't suitable for our case. Maybe moving the code that splits host and port outside of `ParseAddr`is better. How about something like:
```go

func ParseAddr(addr string) (string, int, error) {
	addr = strings.TrimSpace(addr)
	if addr == "" {
		return defaultHost, 0, nil
	}
	host, port, err := splitHostPort(addr)
	if err != nil {
		return "", 0, ihzerrors.NewClientError("", err, hzerrors.ErrInvalidAddress)
	}
	if port < 0 || port > 65535 {
		return "", 0, ihzerrors.NewClientError("port not in valid range", err, hzerrors.ErrInvalidAddress)
	}
	return host, port, nil
}

func splitHostPort(addr string) (host string, port int, err error) {
	idx := strings.LastIndex(addr, ":")
	if idx == -1 {
		return addr, 0, nil
	}
	port, err = strconv.Atoi(addr[idx+1:])
	if err != nil {
		return "", 0, fmt.Errorf("invalid port")
	}
	host, _, err = net.SplitHostPort(addr)
	if err != nil {
		return "", 0, err
	}
	if host == "" {
		host = defaultHost
	}
	return host, port, nil
}
```

Reply:
This `splitHostPort ` would panic when calling with `192.168.1.1:` arg.

Reply:
What's the panic you're getting?
The following works OK without panics:
```go

func splitHostPort(addr string) (host string, port int, err error) {
	idx := strings.LastIndex(addr, ":")
	if idx == -1 {
		return addr, 0, nil
	}
	port, err = strconv.Atoi(addr[idx+1:])
	if err != nil {
		return "", 0, fmt.Errorf("invalid port")
	}
	host, _, err = net.SplitHostPort(addr)
	if err != nil {
		return "", 0, err
	}
	if host == "" {
		host = defaultHost
	}
	return host, port, nil
}

func main() {
	host, port, err := splitHostPort("192.168.1.1:")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(host, port)
}
```

Reply:
It works well, and is better way than checking if address has port separately.